### ISSUE TITLE
Sqflite deadlock hanging on iOS?

### ISSUE THREAD

**Issue by MarkOfSydney** [tekartik/sqflite#248]
Apple rejected my app because Sqflite hangs on accessing the SQL database on an iOS device, blocking the UI in a deadlock, if it calls a background task (like adding initial app db records while splash screen is shown). 

The same app works flawlessly on Android and also on an Xcode simulator or connected iPhone to my Mac! It only hangs on a TestFlight or App Review app.

Grateful for any feedback whether this is a known issue, if a workaround exits, or a solution is forthcoming.

This approach did not work.
https://github.com/flutter/flutter/issues/27844

And none of the below code blocks worked.

import 'package:path_provider/path_provider.dart';
import 'package:sqflite/sqflite.dart';
import 'package:path/path.dart'; // as p; the 'p' is used only if testing iPhone

//  initDb() async {
//    String databasesPath = await getDatabasesPath();
//    String path = join(databasesPath, 'notes.db');
////    await deleteDatabase(path); // just for testing
//    var db = await openDatabase(path, version: 1, onCreate: _onCreate);
//    return db;
//  }

// The code lines below are for testing on the iPhone: NONE WORKED!
//  Future<Database> initDb() async {
//    Directory directory = await getApplicationDocumentsDirectory();
//    String path = p.join(directory.toString(), 'notes.db');
//    Database database =
//        await openDatabase(path, version: 1, onCreate: _onCreate);
//    return database;
//  }

//  Future<Database> initDb() async {
//    //Get path of the directory for android and iOS.
//    var databasesPath = await getDatabasesPath();
//    String path = p.join(databasesPath, 'notes.db');
//    //open/create database at a given path
//    var notesDatabase =
//        await openDatabase(path, version: 1, onCreate: _onCreate);
//    return notesDatabase;
//  }

  // The initDB below uses Path Provider package
  initDb() async {
    Directory documentDirectory = await getApplicationDocumentsDirectory();
    String path = join(documentDirectory.path, "notes.db");
    var ourDb = await openDatabase(path, version: 1, onCreate: _onCreate);
    return ourDb;
  }

**Comment by alextekartik** [comment]
I was not aware of such issue and I'm not an iOS expert. I googled for other having similar rejection (not related to sqflite) like I think you did. Some are flutter related (https://github.com/flutter/flutter/issues/22765). What I would try:
- make sure your app runs on iPad (it seems Apple test on iPad, even iPhone only app)
- call runApp as early as possible (i.e. open the database after, opening the database could be a lengthy operation especially you have an heavy onCreate method), it seems that we stay on the splash screen until then and Apple seems to reject if it is too long (whatever long means...)
- make sure the database dir exists (or just give a relative path - it will be relative to getDatabasesPath() and sqflite will create the dir if needed)

Sorry if I cannot help more. I don't know how you found out that sqflite was hanging (maybe they just the current stack trace at a given time)

**Comment by MarkOfSydney** [comment]
Thanks for the prompt reply. I checked https://github.com/flutter/flutter/issues/22765 and it's precisely my experience to date. So, I removed the reference to Sqflite and the code to open on a non-Sqflite view (no call to database, just static screen), and the app was frozen at the splash screen.

So, it's NOT related to Sqflite; seems like a Flutter/Xcode issue. Thanks again for your prompt reply as it helped me identify what's NOT the problem, and I can focus my efforts elsewhere. All the best.
